calyx/src/passes/compile_sync.rs

```diff
@@ -0,0 +1,380 @@
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use crate::ir::traversal::{Action, Named, VisResult, Visitor};
+use crate::ir::RRC;
+use crate::ir::{self, GetAttributes};
+use crate::{build_assignments, guard, structure};
+use linked_hash_map::LinkedHashMap;
+
+#[derive(Default)]
+/// 1. loop through all control statements under "par" block to find # barriers
+///    needed and # members of each barrier
+/// 2. add all cells and groups needed
+/// 3. loop through all control statements, find the statements with @sync
+///    attribute and replace them with
+///     seq {
+///       <stmt>;
+///       incr_barrier_0_*;
+///       write_barrier_0_*;
+///       wait_*;
+///       restore_*;
+///     }
+///    or
+///     seq {
+///       <stmt>;
+///       incr_barrier_*_*;
+///       write_barrier_*_*;
+///       wait_*;
+///       wait_restore_*;
+///     }
+
+pub struct CompileSync;
+
+/// the structure used to store cells and groups shared by one barrier
+type BarrierMap = LinkedHashMap<u64, (Vec<RRC<ir::Cell>>, Vec<RRC<ir::Group>>)>;
+
+impl Named for CompileSync {
+    fn name() -> &'static str {
+        "compile-sync"
+    }
+
+    fn description() -> &'static str {
+        "Implement barriers for statements marked with @sync attribute"
+    }
+}
+
+fn build_barriers(
+    builder: &mut ir::Builder,
+    s: &mut ir::Control,
+    map: &mut BarrierMap,
+    count: &mut HashMap<u64, u64>,
+) {
+    match s {
+        ir::Control::Seq(seq) => {
+            // go through each control statement
+            // if @sync
+            // see if we already have the set of shared primitives and groups
+            // initialized
+            // True -> generate the inidividual groups and buikld the seq stmt
+            // False -> generate the shared groups, cells
+            //          put the shared groups in the barriermap
+            //          generate the individual groups
+            //          build the seq stmt
+            let mut stmts_new: Vec<ir::Control> = Vec::new();
+            for stmt in seq.stmts.drain(..) {
+                if let Some(n) = stmt.get_attributes().unwrap().get("sync") {
```

Comment:
Consider derefencing `n` right here using `Some(&n)` to avoid all the explicit `*n` down below